google/cloud/aiplatform/jobs.py

```diff
@@ -15,10 +15,214 @@
+            resource_name=batch_prediction_job_name,
+            resource_noun="batchPredictionJobs",
+            project=project,
+            location=location,
+        )
+
+        super().__init__(
+            valid_job_name=valid_batch_prediction_job_name,
+            project=project,
+            location=location,
+            credentials=credentials,
+        )
+
+    def iter_outputs(
+        self, bq_query_limit: Optional[int] = 100
+    ) -> Iterable[Union[storage.Blob, bigquery.job.QueryJob]]:
+        """Returns an Iterable object to traverse the output files, either a list
+        of GCS Blobs or a BigQuery QueryJob depending on the output config set
+        when the BatchPredictionJob was created.
+
+        Args:
+            bq_query_limit: Optional[int] = 100
+                Limit on rows to select from prediction table in BigQuery dataset.
+                Only used when retrieving predictions from a bigquery_destination_prefix.
+                Default is 100.
+
+        Returns:
+            Iterable[Union[storage.Blob, bigquery.job.QueryJob]]:
+                Either a list of GCS Blob objects within the prediction output
+                directory or an iterable QueryJob with predictions.
+
+        Raises:
+            RuntimeError:
+                If BatchPredictionJob is in a JobState other than SUCCEEDED,
+                since outputs cannot be retrieved until the Job has finished.
+            NotImplementedError:
+                If BatchPredictionJob succeeded and output_info does not have a
+                GCS or BQ output provided.
+        """
+
+        job_status = self.status()
+
+        if job_status != JobState.JOB_STATE_SUCCEEDED:
+            raise RuntimeError(
+                f"Cannot read outputs until BatchPredictionJob has succeeded, "
+                f"current status: {job_status}"
+            )
+
+        output_info = self._gca_resource.output_info
+
+        # GCS Destination, return Blobs
+        if output_info.gcs_output_directory not in ("", None):
+            storage_client = storage.Client()
+            blobs = storage_client.list_blobs(output_info.gcs_output_directory)
+            return blobs
+
+        # BigQuery Destination, return QueryJob
+        elif output_info.bigquery_output_dataset not in ("", None):
+            bq_client = bigquery.Client()
+            bigquery.Client
+
+            # Format from service is `bq://projectId.bqDatasetId`
+            bq_dataset = output_info.bigquery_output_dataset
+
+            if bq_dataset.startswith("bq://"):
+                bq_dataset = bq_dataset[5:]
+            if bq_dataset.endswith(("/", ".")):
+                bq_dataset = bq_dataset[:-1]
+
+            # # Split project ID and BQ dataset ID
+            _, bq_dataset_id = bq_dataset.split(".", 1)
+
+            query_limit = f"LIMIT {bq_query_limit}" if bq_query_limit else ""
+            query = f"SELECT * FROM {bq_dataset_id}.predictions {query_limit}"
```

Comment:
Can we get some folks from BigQuery to take a look?  feels like this could be vulnerable to injection attacks?

Reply:
+1 on BigQuery review. 

I thought about the injection attack issue but this method would only succeed if the user had valid credentials to run this query on BQ instance within the same GCP project, but I could be overlooking something. To prevent a mistake, I could validate `bq_query_limit` is actually an `int`

Reply:
`SELECT *` is actually incredibly wasteful. `bq_client.list_rows` is anywhere from 100% (free) to 80% cheaper (BQ Storage API). ([API reference](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html#google.cloud.bigquery.client.Client.list_rows)) It accepts a `max_results` parameter, which can be used instead of a query with a `LIMIT` clause.

Aside: If this was a more complex query. It would be helpful to have some validation here on the dataset ID. Also, I believe you can set a limit using query parameters ([API reference](https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.QueryJobConfig.html#google.cloud.bigquery.job.QueryJobConfig.query_parameters), [docs, with samples](https://cloud.google.com/bigquery/docs/parameterized-queries))

Reply:
Thanks for pointing this out - I've made the change to `list_rows()`.